File: navigators-flutter/lib/src/services/volunteer_service.dart
import 'dart:convert';

import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:http/http.dart' as http;
import 'package:eden_platform_flutter/eden_platform.dart';

/// VolunteerService provides API client for onboarding, leaderboard, and training
/// via ConnectRPC JSON protocol.
///
/// Combines OnboardingService, LeaderboardService, and TrainingService RPCs.
class VolunteerService {
  final String _baseUrl;
  final String? Function() _getAccessToken;

  VolunteerService({
    required String baseUrl,
    required String? Function() getAccessToken,
  })  : _baseUrl = baseUrl,
        _getAccessToken = getAccessToken;

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (_getAccessToken() != null)
          'Authorization': 'Bearer ${_getAccessToken()}',
      };

  Future<Map<String, dynamic>> _post(
      String service, String method, Map<String, dynamic> body) async {
    final url = Uri.parse('$_baseUrl/navigators.v1.$service/$method');
    final response =
        await http.post(url, headers: _headers, body: jsonEncode(body));
    if (response.statusCode != 200) {
      throw Exception(
          '$method failed: ${response.statusCode} ${response.body}');
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  // --- Onboarding ---

  /// Get current user's onboarding status.
  Future<Map<String, dynamic>> getOnboardingStatus() async {
    return _post('OnboardingService', 'GetOnboardingStatus', {});
  }

  /// Acknowledge legal agreement with version string.
  Future<Map<String, dynamic>> acknowledgeLegal(String version) async {
    return _post('OnboardingService', 'AcknowledgeLegal', {
      'version': version,
    });
  }

  /// Complete onboarding (marks user as fully onboarded).
  Future<Map<String, dynamic>> completeOnboarding() async {
    return _post('OnboardingService', 'CompleteOnboarding', {});
  }

  /// Update leaderboard opt-in preference.
  Future<Map<String, dynamic>> updateLeaderboardOptIn(bool optIn) async {
    return _post('OnboardingService', 'UpdateLeaderboardOptIn', {
      'optIn': optIn,
    });
  }

  // --- Leaderboard ---

  /// Get leaderboard entries for a time window (this_week, this_month, all_time).
  Future<Map<String, dynamic>> getLeaderboard(String timeWindow) async {
    return _post('LeaderboardService', 'GetLeaderboard', {
      'timeWindow': timeWindow,
    });
  }

  // --- Training ---

  /// List all training materials.
  Future<List<Map<String, dynamic>>> listTrainingMaterials() async {
    final result =
        await _post('TrainingService', 'ListTrainingMaterials', {});
    return (result['materials'] as List<dynamic>? ?? [])
        .map((m) => m as Map<String, dynamic>)
        .toList();
  }

  /// Get a presigned download URL for a training material.
  Future<String> getTrainingDownloadUrl(String materialId) async {
    final result = await _post('TrainingService', 'GetTrainingDownloadUrl', {
      'materialId': materialId,
    });
    return result['url'] as String? ?? result['presignedUrl'] as String? ?? '';
  }

  /// Get a presigned PUT URL for direct upload of a training material to MinIO.
  /// Returns a map with keys: presignedUrl, storageKey, expiresInSeconds.
  Future<Map<String, dynamic>> getTrainingUploadUrl({
    required String filename,
    required String contentType,
  }) async {
    return _post('TrainingService', 'GetTrainingUploadUrl', {
      'filename': filename,
      'contentType': contentType,
    });
  }

  /// Create a training material metadata record pointing at an already-uploaded object.
  Future<Map<String, dynamic>> createTrainingMaterial({
    required String title,
    required String description,
    required String contentUrl,
    required int sortOrder,
  }) async {
    return _post('TrainingService', 'CreateTrainingMaterial', {
      'title': title,
      'description': description,
      'contentUrl': contentUrl,
      'sortOrder': sortOrder,
    });
  }

  /// Update an existing training material's metadata.
  Future<Map<String, dynamic>> updateTrainingMaterial({
    required String id,
    required String title,
    required String description,
    required int sortOrder,
    required bool isPublished,
  }) async {
    return _post('TrainingService', 'UpdateTrainingMaterial', {
      'id': id,
      'title': title,
      'description': description,
      'sortOrder': sortOrder,
      'isPublished': isPublished,
    });
  }

  /// Soft-delete a training material (server flips is_published=false).
  Future<void> deleteTrainingMaterial(String id) async {
    await _post('TrainingService', 'DeleteTrainingMaterial', {'id': id});
  }
}

/// Riverpod provider for VolunteerService.
final volunteerServiceProvider = Provider<VolunteerService>((ref) {
  final auth = ref.watch(authProvider);
  const envUrl = String.fromEnvironment('API_BASE_URL', defaultValue: '');
  final baseUrl = envUrl.isNotEmpty ? envUrl : 'http://localhost:8080';
  return VolunteerService(
    baseUrl: baseUrl,
    getAccessToken: () => auth.accessToken,
  );
});

File: navigators-go/internal/navigators/volunteer_service.go
package navigators

import (
	"context"
	"fmt"
	"net/url"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/minio/minio-go/v7"

	"navigators-go/internal/db"
)

// VolunteerService provides onboarding, leaderboard, and training material operations.
type VolunteerService struct {
	queries        *db.Queries
	pool           *pgxpool.Pool
	minioClient    *minio.Client
	trainingBucket string
}

// NewVolunteerService creates a new VolunteerService.
func NewVolunteerService(queries *db.Queries, pool *pgxpool.Pool, minioClient *minio.Client, trainingBucket string) *VolunteerService {
	return &VolunteerService{
		queries:        queries,
		pool:           pool,
		minioClient:    minioClient,
		trainingBucket: trainingBucket,
	}
}

// GetOrCreateProfile returns an existing profile or creates a new one.
func (s *VolunteerService) GetOrCreateProfile(ctx context.Context, companyID, userID uuid.UUID) (*db.NavigatorProfile, error) {
	profile, err := s.queries.UpsertNavigatorProfile(ctx, db.UpsertNavigatorProfileParams{
		UserID:    userID,
		CompanyID: companyID,
	})
	if err != nil {
		return nil, fmt.Errorf("upsert navigator profile: %w", err)
	}
	return &profile, nil
}

// AcknowledgeLegal records a legal acknowledgment with version string for audit.
func (s *VolunteerService) AcknowledgeLegal(ctx context.Context, userID uuid.UUID, version string) error {
	if err := s.queries.UpdateLegalAcknowledgment(ctx, db.UpdateLegalAcknowledgmentParams{
		UserID:                     userID,
		LegalAcknowledgmentVersion: &version,
	}); err != nil {
		return fmt.Errorf("update legal acknowledgment: %w", err)
	}
	return nil
}

// CompleteOnboarding marks onboarding as complete, gating app access.
func (s *VolunteerService) CompleteOnboarding(ctx context.Context, userID uuid.UUID) (*db.NavigatorProfile, error) {
	profile, err := s.queries.CompleteOnboarding(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("complete onboarding: %w", err)
	}
	return &profile, nil
}

// UpdateLeaderboardOptIn toggles leaderboard participation.
func (s *VolunteerService) UpdateLeaderboardOptIn(ctx context.Context, userID uuid.UUID, optIn bool) error {
	if err := s.queries.UpdateLeaderboardOptIn(ctx, db.UpdateLeaderboardOptInParams{
		UserID:           userID,
		LeaderboardOptIn: optIn,
	}); err != nil {
		return fmt.Errorf("update leaderboard opt-in: %w", err)
	}
	return nil
}

// GetLeaderboard returns aggregated metrics for opted-in navigators.
func (s *VolunteerService) GetLeaderboard(ctx context.Context, companyID uuid.UUID, timeWindow string) ([]db.GetLeaderboardRow, error) {
	var since time.Time
	switch timeWindow {
	case "week":
		since = time.Now().UTC().AddDate(0, 0, -7)
	case "month":
		since = time.Now().UTC().AddDate(0, 0, -30)
	case "all_time":
		since = time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
	default:
		return nil, fmt.Errorf("invalid time_window: %s (must be week, month, or all_time)", timeWindow)
	}

	rows, err := s.queries.GetLeaderboard(ctx, db.GetLeaderboardParams{
		CompanyID:   companyID,
		CheckedInAt: since,
	})
	if err != nil {
		return nil, fmt.Errorf("get leaderboard: %w", err)
	}
	return rows, nil
}

// ListTrainingMaterials returns published training materials for a company.
func (s *VolunteerService) ListTrainingMaterials(ctx context.Context, companyID uuid.UUID) ([]db.TrainingMaterial, error) {
	materials, err := s.queries.ListTrainingMaterials(ctx, companyID)
	if err != nil {
		return nil, fmt.Errorf("list training materials: %w", err)
	}
	return materials, nil
}

// CreateTrainingMaterial creates a new training material entry.
func (s *VolunteerService) CreateTrainingMaterial(ctx context.Context, companyID, userID uuid.UUID, title, description, contentURL string, sortOrder int32) (*db.TrainingMaterial, error) {
	material, err := s.queries.CreateTrainingMaterial(ctx, db.CreateTrainingMaterialParams{
		CompanyID:   companyID,
		Title:       title,
		Description: description,
		ContentUrl:  contentURL,
		SortOrder:   sortOrder,
		CreatedBy:   userID,
	})
	if err != nil {
		return nil, fmt.Errorf("create training material: %w", err)
	}
	return &material, nil
}

// UpdateTrainingMaterial updates mutable fields on a training material.
func (s *VolunteerService) UpdateTrainingMaterial(ctx context.Context, companyID, id uuid.UUID, title, description string, sortOrder int32, isPublished bool) (*db.TrainingMaterial, error) {
	material, err := s.queries.UpdateTrainingMaterial(ctx, db.UpdateTrainingMaterialParams{
		ID:          id,
		CompanyID:   companyID,
		Title:       title,
		Description: description,
		SortOrder:   sortOrder,
		IsPublished: isPublished,
	})
	if err != nil {
		return nil, fmt.Errorf("update training material: %w", err)
	}
	return &material, nil
}

// DeleteTrainingMaterial soft-deletes a training material (is_published=false).
// The row is retained so pull sync can propagate the removal to clients.
func (s *VolunteerService) DeleteTrainingMaterial(ctx context.Context, companyID, id uuid.UUID) error {
	if err := s.queries.SoftDeleteTrainingMaterial(ctx, db.SoftDeleteTrainingMaterialParams{
		ID:        id,
		CompanyID: companyID,
	}); err != nil {
		return fmt.Errorf("soft delete training material: %w", err)
	}
	return nil
}

// GetTrainingUploadURL returns a 15-minute presigned PUT URL and a server-generated
// storage key. The client uploads bytes directly to MinIO, then passes the same
// storageKey to CreateTrainingMaterial as content_url.
func (s *VolunteerService) GetTrainingUploadURL(ctx context.Context, filename, _contentType string) (string, string, int32, error) {
	if filename == "" {
		return "", "", 0, fmt.Errorf("filename is required")
	}
	storageKey := uuid.NewString() + "/" + filename
	u, err := s.minioClient.PresignedPutObject(ctx, s.trainingBucket, storageKey, 15*time.Minute)
	if err != nil {
		return "", "", 0, fmt.Errorf("presign training upload: %w", err)
	}
	return u.String(), storageKey, int32(15 * 60), nil
}

// GetTrainingDownloadURL returns a presigned download URL for a training material.
func (s *VolunteerService) GetTrainingDownloadURL(ctx context.Context, materialID uuid.UUID) (string, error) {
	material, err := s.queries.GetTrainingMaterial(ctx, materialID)
	if err != nil {
		return "", fmt.Errorf("get training material: %w", err)
	}

	// Generate a 15-minute presigned URL for the MinIO object
	reqParams := make(url.Values)
	presignedURL, err := s.minioClient.PresignedGetObject(ctx, s.trainingBucket, material.ContentUrl, 15*time.Minute, reqParams)
	if err != nil {
		return "", fmt.Errorf("generate presigned URL: %w", err)
	}

	return presignedURL.String(), nil
}
